torchvision/datasets/utils.py

```diff
@@ -118,7 +122,7 @@ def download_url(
     root = os.path.expanduser(root)
     if not filename:
         filename = os.path.basename(url)
-    fpath = os.path.join(root, filename)
+    fpath = os.fspath(os.path.join(root, filename))
```

Comment:
Hm, is `os.fspath` what we need here? I think it should suffice to do something like

```suggestion
    fpath = os.path.join(str(root), filename)
```

which would make sure to convert `root` to a `str` object, in case it was a `pathlib.Path` object. But we'll confirm once we write tests.

Reply:
I did other patches using str(path) but I believe os.fspath(path) is the right way to do this:

https://discuss.python.org/t/make-pathlib-extensible/3428/79

Also the issue number was incorrect. This PR is supposed to partially address #8120

Reply:
There is more discussion about str() vs. fspath here:

https://peps.python.org/pep-0519/

That proposal appears to be final and the relevant quote from there is this:

One issue in converting path objects to strings comes from the fact that the only generic way to get a string representation of the path was to pass the object to str(). This can pose a problem when done blindly as nearly all Python objects have some string representation whether they are a path or not, e.g. str(None) will give a result that builtins.open() [[5]](https://peps.python.org/pep-0519/#builtins-open) will happily use to create a new file.

Reply:
Sounds good, I hadn't thought about the `str(None)` issue. Let's use `fspath` then, thank you for the investigation!